spec/design.md

```diff
@@ -202,19 +218,22 @@ where:
 
 ### Generics
 
-Calling `defineGeneric()` defines a new generic. It has the signature:
+Calling `newGeneric()` defines a new generic. It has the signature:
+
 ```{R}
-function(name, FUN, signature) { }
+newGeneric(name, FUN, signature)
 ```
+
 The `signature` would default to all of the arguments in
-`formals(FUN)`. The body of `FUN` would resemble S3 and S4
+`formals(FUN)[1]`. The body of `FUN` would resemble S3 and S4
 generics. It might just call `UseMethod()`.
 
 ### Methods
 
-Calling `defineMethod()` defines a new method, with the signature:
+Calling `newMethod()` defines a new method, with the signature:
+
 ```{R}
-function(generic, signature, FUN) { }
+newMethod(generic, FUN, signature)
```

Comment:
Moved signature to end for consistency with `newGeneric`. Is it inconsistent that this object doesn't contain the class name?

Reply:
I see the need for consistency. This may just be my S4 lens, but it's been useful having the signature argument before the implementation, since it's a lot easier to see. Then again, people may just use the syntactic sugar instead.

Reply:
Also, the two `signature` arguments have different meanings. For generics, it is a list of the argument names on which to dispatch. For methods, it is a vector with a class name for every argument in the generic signature. Maybe we need different names? Like `newGeneric()` could have `signatureArgs` instead of `signature`.

Reply:
Oh true. I was also wondering about `newGeneric()` — if we have the name of the generic, then we could actually generate the body of the function, but then what would `fun` be?

Reply:
Maybe the `newMethod()` argument name could contain `class`?

Reply:
The only reason I put `FUN` there was to get the formals. Seems more natural to pass a function than a call to `alist()`. I agree the body is redundant, just like S4.

Reply:
The `class` suggestion seems reasonable. Please commit a proposal.

Reply:
I'm going to turn it back to `signature` and merge this PR. Then we can discuss the interface in more detail in #29.